pyiceberg/avro/resolver.py

```diff
@@ -233,7 +255,107 @@ def skip(self, decoder: BinaryDecoder) -> None:
         pass
 
 
-class SchemaResolver(PrimitiveWithPartnerVisitor[IcebergType, Reader]):
+class WriteSchemaResolver(PrimitiveWithPartnerVisitor[IcebergType, Writer]):
+    def schema(self, schema: Schema, expected_schema: Optional[IcebergType], result: Writer) -> Writer:
+        return result
+
+    def struct(self, struct: StructType, provided_struct: Optional[IcebergType], field_writers: List[Writer]) -> Writer:
+        if not isinstance(provided_struct, StructType):
+            raise ResolveError(f"File/write schema are not aligned for struct, got {provided_struct}")
+
+        provided_struct_positions: Dict[int, int] = {field.field_id: pos for pos, field in enumerate(provided_struct.fields)}
+
+        results: List[Tuple[Optional[int], Writer]] = []
+        iter(field_writers)
+
+        for pos, write_field in enumerate(struct.fields):
+            if write_field.field_id in provided_struct_positions:
+                results.append((provided_struct_positions[write_field.field_id], field_writers[pos]))
+            else:
+                # There is a default value
+                if isinstance(write_field, NestedField) and write_field.write_default is not None:
+                    # The field is not in the record, but there is a write default value
+                    default_writer = DefaultWriter(
+                        writer=visit(write_field.field_type, CONSTRUCT_WRITER_VISITOR), value=write_field.write_default
+                    )
+                    results.append((None, default_writer))
+                elif write_field.required:
+                    raise ValueError(f"Field is required, and there is no write default: {write_field}")
+                else:
+                    results.append((pos, NoneWriter()))
```

Comment:
I don't think this is correct if `NoneWriter` writes data. There is a field in the data schema that is not present in the file schema. It should just be skipped in that case rather than writing a null byte, right?

Reply:
Yes, this one can go completely. This didn't show up since I was testing exclusively with V1 tables (so there are no deprecated fields)